plugins/csi/client.go

```diff
@@ -254,6 +254,36 @@ func (c *client) ControllerPublishVolume(ctx context.Context, req *ControllerPub
 	}, nil
 }
 
+func (c *client) ControllerUnpublishVolume(ctx context.Context, req *ControllerUnpublishVolumeRequest) (*ControllerUnpublishVolumeResponse, error) {
+	if c == nil {
+		return nil, fmt.Errorf("Client not initialized")
+	}
+	if c.controllerClient == nil {
+		return nil, fmt.Errorf("controllerClient not initialized")
+	}
+	if req.VolumeID == "" {
+		return nil, fmt.Errorf("missing VolumeID")
+	}
+
+	if req.NodeID == "" {
+		// the spec allows this but it would unpublish the
+		// volume from all nodes
+		return nil, fmt.Errorf("missing NodeID")
+	}
+
+	upbrequest := &csipbv1.ControllerUnpublishVolumeRequest{
```

Comment:
use `req.ToCSIRequest()`? Could also implement validation on the struct too maybe but nbd

Reply:
It looks like we typically validate client RPCs in the endpoints and not in the request structs. But taking a second look at `client/structs.go`, I'm wondering if all the request structs in this file should embed `structs.QueryOptions` and all the responses `structs.QueryMeta`?

Reply:
Probably not - `QueryOptions` and `QueryMeta` are only really for HTTP APIs rather than internal RPCs (and don't make a lot of sense when Raft isn't involved)

Reply:
Gotcha.

Reply:
I managed to break everything when swapping out for `ToCSIRequest`, as that method is defined upstream. I need to implement the `ToCSIRepresentation` found in `./plugins/csi/plugins.go` first. Looks like that's where we did the validation for the node publish too. Lemme do that and that'll clean this up nicely.